field/FieldCreate.F90

```diff
@@ -84,18 +85,45 @@ subroutine field_empty_complete( field, &
       character(len=*), optional, intent(in) :: long_name
       integer, optional, intent(out) :: rc
 
-      integer :: status
       type(LU_Bound), allocatable :: bounds(:)
       type(ESMF_Info) :: field_info
       type(VerticalStaggerLoc) :: vert_staggerloc_
+      integer, allocatable :: grid_to_field_map(:)
+      type(ESMF_Geom) :: geom
+      real(kind=ESMF_KIND_R4), allocatable :: farray(:)
+      integer :: dim_count, idim, status
+
+      if (present(gridToFieldMap)) then
+         grid_to_field_map = gridToFieldMap
+      else
+         call ESMF_FieldGet(field, geom=geom, _RC)
+         call ESMF_GeomGet(geom, dimCount=dim_count, _RC)
+         allocate(grid_to_field_map(dim_count), source=[(idim, idim=1,dim_count)])
+      end if
 
       bounds = make_bounds(num_levels=num_levels, ungridded_dims=ungridded_dims)
-      call ESMF_FieldEmptyComplete(field, typekind=typekind, &
-           gridToFieldMap=gridToFieldMap, &
-           ungriddedLBound=bounds%lower, ungriddedUBound=bounds%upper, _RC)
+      if (all(grid_to_field_map == 0)) then
+         _ASSERT(typekind==ESMF_TYPEKIND_R4, "only r4 arrays supported for vert only fields")
+         if (present(ungridded_dims)) then
+            _ASSERT(ungridded_dims%get_num_ungridded() == 0, "ungridded dims not supported for vert only fields")
+         end if
+         allocate(farray(num_levels), source=MAPL_UNDEFINED_REAL)
+         call ESMF_FieldEmptyComplete( &
+              field, &
+              farray=farray, &
+              indexFlag=ESMF_INDEX_DELOCAL, &
+              datacopyFlag=ESMF_DATACOPY_VALUE, &
+              gridToFieldMap=grid_to_field_map, &
+              ungriddedLBound=bounds%lower, &
+              ungriddedUBound=bounds%upper, &
+              _RC)
+      else
+         call ESMF_FieldEmptyComplete(field, typekind=typekind, &
+              gridToFieldMap=gridToFieldMap, &
+              ungriddedLBound=bounds%lower, ungriddedUBound=bounds%upper, _RC)
```

Comment:
I do not like this divergence in the creation arguments.   In particular, it is not at all obvious why vert only has DATACOPY and INDEX arguments.  Nor why it is returning an farray.

Reply:
I addressed the issue some in the description of the PR. Essentially, there seems to be a bug in the the type kind interface of `ESMF_FieldEmptyComplete`, so we use the `farray` interface. A bug report was filed, and once it is fixed, this extra bit of code goes away.